### DISCUSSION TITLE
Batch Job & onFailure(Job job). How do they work?

### DISCUSSION THREAD

**Discussion by VN11733** [jobrunr/jobrunr#1114] [Q&A] [answered]
### Issue Description
We're facing with unpredictable behaviour from Batch Job as well as `onFailure` function (Code example is provided below):
- batches doesn't work as they written
- onFailure doesn't trigger when exception was thrown within the chain.

### Case Description
Let's say we have an Quarkus app with two identical postgres tables: `citizen` and `embassy`. I created a chain with next jobs:
1. Create table citizen
2. Flush table citizen
3. Create table embassy
4. Flush table embassy
5. Init Batch for inserting data to citizen and embassy
6. On failure, print my error.

In batch, I'd like to do another chain:
1. Insert data to citizen
2. Insert data to embassy
3. On failure, print my error.

Code Example:

```java
@ApplicationScoped
public class CitizenJobProcessor {

    private static final Logger logger = LoggerFactory.getLogger(CitizenJobProcessor.class);
    private static final int CITIZENS_COUNT = 10000;
    private static final int CITIZENS_BATCH_SIZE = 100;

    public CitizenJobProcessor() {}

    @Inject
    EmbassyJobProcessor embassyJobProcessor;

    @Inject
    JobScheduler jobScheduler;

    @Inject
    @Named(Datasources.MY_DATASOURCE)
    SimpleDAO simpleDAO;

    public void runCitizensBatch() {
        jobScheduler
                .enqueue(this::runCreateTable)
                .continueWith(this::deleteCitizens)
                .continueWith(embassyJobProcessor::runCreateTable)
                .continueWith(embassyJobProcessor::runDeleteAll)
                .continueWith(this::initBatch)
                .onFailure(this::printProcessError);
    }

    @Job(name = "Print CitizenJobProcessor::runCitizensProcess Error", retries = 0)
    public void printProcessError() {
        logger.error("Something went wrong during processing pipeline");
    }

    @Job(name = "Create Citizen Table", retries = 0)
    @Transactional
    public void runCreateTable() throws SQLException {
        logger.info("Creating Table my_schema.citizen");
        simpleDAO.executePrepared("""
                create table if not exists my_schema.citizen
                (
                id serial primary key,
                first_name text not null,
                last_name text not null,
                birth_date date not null,
                gender text,
                city_of_residence text,
                nationality text,
                phone_number text,
                email text,
                insert_timestamp timestamp not null default now()
                )
                """);
        logger.info("Table my_schema.citizen is created.");
    }

    @Job(name = "Init Batch for Inserting Citizens", retries = 0)
    public void initBatch() {
        logger.info("Initializing Batch for inserting citizens data.");
        jobScheduler.startBatch(this::generateBatch).onFailure(this::printBatchError);
    }

    @Job(name = "Generate Batch for Inserting Citizens", retries = 0)
    public void generateBatch() {
        logger.info("Generating batch for inserting citizens...");
        final List<Citizen> citizens = CitizenGenerator.generateCitizens(CITIZENS_COUNT);
        List<Citizen> citizensBatch = new ArrayList<>(CITIZENS_BATCH_SIZE);
        int counter = 0;
        for (int idx = 0; idx < CITIZENS_COUNT; idx++) {
            citizensBatch.add(citizens.get(idx));
            counter++;
            if (counter != CITIZENS_BATCH_SIZE) {
                continue;
            }
            jobScheduler.enqueue(() -> insertCitizens(citizensBatch))
                    .continueWith(() -> embassyJobProcessor.runInsertEmbassy(citizensBatch))
                    .onFailure(this::printBatchError);
            citizensBatch.clear();
            counter = 0;
        }
        logger.info("Citizen Batch is generated.");
    }

    @Job(name = "Delete Citizens Data", retries = 0)
    @Transactional
    public void deleteCitizens() throws SQLException {
        logger.info("Flushing data from indigo.ilia_citizen...");
        simpleDAO.executePrepared("TRUNCATE my_schema.citizen");
        logger.info("Data from my_schema.citizen is flushed.");
    }

    @Job(name = "Insert Citizens", retries = 0)
    @Transactional
    public void insertCitizens(List<Citizen> citizens) throws SQLException {
        final Logger logger = LoggerFactory.getLogger("CitizenJobProcessor::insertCitizens");
        logger.info("Adding citizens from {} to {}", citizens.get(0).phoneNumber(), citizens.get(citizens.size() - 1).phoneNumber());
        final String prefix = """
                insert into my_schema.citizen (
                first_name,
                last_name,
                birth_date,
                gender,
                city_of_residence,
                nationality,
                phone_number,
                email,
                insert_timestamp
                )
                """;
        final String values = citizens.stream().map(citizen -> """
                ('%s', '%s', '%s', %s, %s, %s, %s, %s, %s)
                """.formatted(
                citizen.firstName(),
                citizen.lastName(),
                citizen.birthDate().toString(),
                getQuotedOrNull(citizen.gender()),
                getQuotedOrNull(citizen.cityOfResidence()),
                getQuotedOrNull(citizen.nationality()),
                getQuotedOrNull(citizen.phoneNumber()),
                getQuotedOrNull(citizen.email()),
                "NOW()"
        )).collect(Collectors.joining(Constants.COMMA));
        simpleDAO.executePrepared("%s values %s".formatted(prefix, values));
        logger.info("Citizens were added");
    }

    @Job(name = "Print Batch Processing Error", retries = 0)
    public void printBatchError() {
        logger.error("Something went wrong during batch processing");
    }
}
```
In the written code you can see that we want to insert data batch (100 items) to 1st table, then to 2nd one, but in practice JobRunr runs `() -> insertCitizens(citizensBatch)` 100 times (10_000 : 100), then runs `() -> embassyJobProcessor.runInsertEmbassy(citizensBatch)` another 100 times, which is wrong.
Another problem is whenever I inject a bug on purpose to trigger `onFailure`, it doesn't. Shouldn't it be called in case of exception was thrown or it is for different things?

P.S. I've read documentation, but it doesn't provide answers for my questions.

**Comment by auloin** [comment]
Hi @VN11733,

When `continueWith(continuationJob)` is used, the `continuationJob` is executed after the previous job in the chain has succeeded. E.g.,
```
jobScheduler
                .enqueue(this::runCreateTable)
                .continueWith(this::deleteCitizens)
                .continueWith(embassyJobProcessor::runCreateTable)
```
`this::deleteCitizens` runs after `this::runCreateTable` has successfully run and `embassyJobProcessor::runCreateTable` runs after `this::deleteCitizens` has successfully run.

This is why `() -> embassyJobProcessor.runInsertEmbassy(citizensBatch)` runs after `() -> insertCitizens(citizensBatch)`. If you don't want this chain you can enqueue them separately.

When `onFailure(continuationJob)` is used, the `continuationJob` is executed after the previous job in the chain has failed. E.g.,

```
jobScheduler.enqueue(() -> insertCitizens(citizensBatch))
                    .continueWith(() -> embassyJobProcessor.runInsertEmbassy(citizensBatch))
                    .onFailure(this::printBatchError);
```

`this::printBatchError` runs only if `() -> embassyJobProcessor.runInsertEmbassy(citizensBatch)` fails.

> Another problem is whenever I inject a bug on purpose to trigger `onFailure`, it doesn't. Shouldn't it be called in case of exception was thrown or it is for different things?

So if you're introducing a bug to make `() -> insertCitizens(citizensBatch)` fail, `this::printBatchError` is not going to run.

If you need to run a `onFailure` job at each step, you can use `continueWith(onSuccessContinuationJob, onFailureContinuationJob)`. Note that `onFailureContinuationJob`s end the chain, meaning if the previous job fails only it's `onFailureContinuationJob` is executed, all the remaining jobs in the chain (`onSuccessContinuationJob` or `onFailureContinuationJob`) are set to `NOT_APPLICABLE` and won't be executed.

Does this help? Regarding the comments on the BatchJobs, in my view JobRunr is doing exactly as instructed. Could you elaborate on why you think it's not the case?

**Reply by VN11733** [reply]
Hi @auloin,

First, I'd like to say thank you a lot for your answer, you've answered most of it and you're right with 1st part (I don't have any questions regarding its work), but I still have some questions.

1. If something went wrong during `enqueue(job)`, how can I trigger `onFailure(job)` in that case? Should I start with `continueWith(job, failJob)` instead of `enqueue(job)` in that case?
2. I have an issue with next code:
```java
jobScheduler.enqueue(() -> insertCitizens(citizensBatch))
                    .continueWith(() -> embassyJobProcessor.runInsertEmbassy(citizensBatch))
                    .onFailure(this::printBatchError);
```
It has to insert 100 to citizen and same 100 to embassy right after it and like that other 99 times, but instead it inserts 100 * 100 to citizen and then 100 * 100 to embassy. This behaviour is not obvious from the code: by the code it is supposed to work as I described earlier.

**Reply by auloin** [reply, marked as answer]
> 1. If something went wrong during `enqueue(job)`, how can I trigger `onFailure(job)` in that case? Should I start with `continueWith(job, failJob)` instead of `enqueue(job)` in that case?

The chain starts by `enqueue(job)` (or any `jobScheduler` method that returns a `JobProId`). And continues with another job.

```
jobScheduler.enqueue(jobA)
                    .continueWith(jobBToRunIfJobASucceeds, jobCToRunIfJobAFails)
                    .continueWith(jobDToRunIfJobBSucceeds, jobEToRunIfJobBFails)
                   //... rest of the chain
;
```
This first enqueues an initial job `jobA` if it succeeds `jobBToRunIfJobASucceeds` runs otherwise `jobCToRunIfJobAFails` runs and the chains ends. If  `jobBToRunIfJobASucceeds` can run (so `jobA` succeeds);  if it succeeds, `jobDToRunIfJobBSucceeds` runs, otherwise `jobEToRunIfJobBFails` runs and the chain ends.

> 2. I have an issue with next code:
> 
> ```java
> jobScheduler.enqueue(() -> insertCitizens(citizensBatch))
>                     .continueWith(() -> embassyJobProcessor.runInsertEmbassy(citizensBatch))
>                     .onFailure(this::printBatchError);
> ```
> 
> It has to insert 100 to citizen and same 100 to embassy right after it and like that other 99 times, but instead it inserts 100 * 100 to citizen and then 100 * 100 to embassy. This behaviour is not obvious from the code: by the code it is supposed to work as I described earlier.

JobRunr processes jobs in the order they are enqueued. The jobs for inserting in the `citizens` table are enqueued first, the jobs for inserting in the `embassy` table are only enqueued once the job they need to wait succeeds. 

From how you describe it, it sounds like you want the operation to be transactional, but chaining jobs doesn't guarantee this.

**Reply by VN11733** [reply]
1. Thank you for an explanation, it helped a lot. I'd like to ask you to provide this info to the documentation, if possible or rephrase the current, because it is not obvious for now, but your explanation is.
2. Oh, so you're saying technically they were run one after another, but it doesn't mean it'll run *right* after previous job. I get it now. I mark your last comment as answer. 

Thanks again for your explanations.
